torch/utils/_python_dispatch.py

```diff
@@ -147,6 +147,10 @@ def __init__(self):
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         raise NotImplementedError()
 
+    @classmethod
+    def push(cls, *args, **kwargs):
+        return push_torch_dispatch_mode(functools.partial(cls, *args, **kwargs))
```

Comment:
Add test in test_python_dispatch?

Reply:
Actually, I was thinking of gussying up the API even more. What if this worked:

```
with MyMode(blah, blah):
  ...
```

The main missing ingredient is that the mode constructor needs to record the current mode to its `inner` (instead of getting it mutated by the push function, which is currently what's happening).

Reply:
Yeah this would be cool!

Reply:
cc @samdow 

What I'm vacillating about is how exactly this should be implemented. An obvious way to do it is

```
class ModeBase:
    def __init__(self):
        self.inner = _get_mode()
```

but I'm afraid people will forget to call `super().__init__` and then inner won't be set (maybe we can detect this happened with hasattr). Alternately, we can use wrap_init in the metaclass and that also makes easy to accept an `inner` kwarg to override it, but it's more implicit.

Reply:
> I'm afraid people will forget to call super().__init__ and then inner won't be set (maybe we can detect this happened with hasattr)

This should be caught during the first call to torch dispatch. We could add a clearer error message with the hasattr check since we have a guess as to what the user did wrong (in general, they shouldn't have to think about inner that much)

> Alternately, we can use wrap_init in the metaclass and that also makes easy to accept an inner kwarg to override it, but it's more implicit.

IMO, this feel nicer since my guess is that most people will want the implicit behavior of just using whatever mode is currently inner. Also since we're already using the metaclass to wrap torch dispatch and init, it fits with the rest of the design. I guess the only question is the flip of the current question: in this version, what happens if a user calls `super().__init__`? Does wrap_init get called twice (maybe not the end of the world)?

Reply:
If we go the meta class route I would just not put anything in the super init. I guess there's a weird problem which is that the metaclass will try to set inner on every super() call, that's kind of weird, not sure if there's a good way around it

Reply:
I agree setting inner on every super call is kind of weird and we could do something like with the hasattr if we really wanted to only set it once. I still think it's better than having an error that makes the user have to know about inner even in the basic case